Add more Tuya Vacuum sensors and select entities and send command service

dougiteixeira opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
* The implementation of the pause vacuum command does not work because it sends True instead of False.
* Implemented the service send command to vacuum.
* Included Clean Area and Clean Time sensors and Status Tuya diagnostic sensor (disabled by default) for vacuum.
![image](https://user-images.githubusercontent.com/31328123/144342467-cc6e2d3d-dde9-4483-9fe3-238af619871c.png) ![image](https://user-images.githubusercontent.com/31328123/144342488-e0e4930d-64ff-44c8-bc77-a4b7f1522ecb.png)

* Included selects Mode, Suction and Direction control for vacuum.
![image](https://user-images.githubusercontent.com/31328123/144342545-ff098378-dcf4-49d4-973e-97ca7fa96639.png)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

dougiteixeira committed:
Adjusted comand PAUSE and Add service send comand

dougiteixeira committed:
Add MODE, DIRECTION CONTROL and SUCTION Vacuum

dougiteixeira committed:
Add CLEAN AREA, TIME and STATUS TUYA Vacuum

dougiteixeira committed:
Add CLEAN AREA, CLEAN TIME and DIRECTION CONTROL

dougiteixeira committed:
Update Sensor STATUS TUYA Vacuum

dougiteixeira committed:
Update select vacuum

dougiteixeira committed:
Update sensor.py

dougiteixeira committed:
Update sensor.py

dougiteixeira committed:
Update const.py

dougiteixeira commented:
There's no breaking change, I checked the box by mistake.

frenck commented on homeassistant/components/tuya/select.py:
```diff
@@ -218,6 +218,26 @@
+            key=DPCode.SUCTION,
+            name="Suction",
+            entity_category=ENTITY_CATEGORY_CONFIG,
+            entity_registry_enabled_default=False,
```
Why is this disabled by default?

dougiteixeira replied:
As there is an option in the vacuum UI I preferred to leave it disabled and the user to enable it if you want to use it in automations. Do you think it should be enabled?

frenck replied:
So if this is already part of the vacuum entity, why add it here again?

frenck commented on homeassistant/components/tuya/select.py:
```diff
@@ -218,6 +218,26 @@
+        SelectEntityDescription(
+            key=DPCode.SUCTION,
+            name="Suction",
+            entity_category=ENTITY_CATEGORY_CONFIG,
```
The Tuya integration has moved to use StrEnums for this.

```suggestion
            entity_category=EntityCategory.CONFIG,
```

frenck commented on homeassistant/components/tuya/select.py:
```diff
@@ -218,6 +218,26 @@
+            key=DPCode.DIRECTION_CONTROL,
+            name="Direction Control",
+            entity_category=ENTITY_CATEGORY_CONFIG,
+            entity_registry_enabled_default=False,
```
Why is this disabled by default?

frenck commented on homeassistant/components/tuya/select.py:
```diff
@@ -218,6 +218,26 @@
+        SelectEntityDescription(
+            key=DPCode.MODE,
+            name="Mode",
+            entity_category=ENTITY_CATEGORY_CONFIG,
```
```suggestion
            entity_category=EntityCategory.CONFIG,
```

frenck commented on homeassistant/components/tuya/select.py:
```diff
@@ -218,6 +218,26 @@
+        SelectEntityDescription(
+            key=DPCode.DIRECTION_CONTROL,
+            name="Direction Control",
+            entity_category=ENTITY_CATEGORY_CONFIG,
```
```suggestion
            entity_category=EntityCategory.CONFIG,
```

frenck commented on homeassistant/components/tuya/sensor.py:
```diff
@@ -594,6 +594,24 @@ class TuyaSensorEntityDescription(SensorEntityDescription):
+        TuyaSensorEntityDescription(
+            key=DPCode.STATUS,
+            name="Status Tuya",
+            entity_category=ENTITY_CATEGORY_DIAGNOSTIC,
```
```suggestion
            entity_category=EntityCategory.DIAGNOSTIC,
```

frenck commented on homeassistant/components/tuya/sensor.py:
```diff
@@ -594,6 +594,24 @@ class TuyaSensorEntityDescription(SensorEntityDescription):
+            name="Clean time",
+        ),
+        TuyaSensorEntityDescription(
+            key=DPCode.STATUS,
```
This is the vacuum state, right? Why introduce that as a separate sensor?

frenck commented on homeassistant/components/tuya/vacuum.py:
```diff
@@ -171,7 +171,7 @@ def stop(self, **kwargs: Any) -> None:
     def pause(self, **kwargs: Any) -> None:
         """Pause the device."""
-        self._send_command([{"code": DPCode.POWER_GO, "value": True}])
+        self._send_command([{"code": DPCode.POWER_GO, "value": False}])
```
This looks like a bug fix, which should be a separate PR.

frenck commented on homeassistant/components/tuya/vacuum.py:
```diff
@@ -184,3 +184,7 @@ def locate(self, **kwargs: Any) -> None:
         """Set fan speed."""
         self._send_command([{"code": DPCode.SUCTION, "value": fan_speed}])
+
+    def send_command(self, command, params, **kwargs) -> None:
+        """Set raw command."""
+        self._send_command(command)
```
What does this do?

frenck commented on homeassistant/components/tuya/sensor.py:
```diff
@@ -594,6 +594,24 @@ class TuyaSensorEntityDescription(SensorEntityDescription):
+    "sd": (
+        TuyaSensorEntityDescription(
+            key=DPCode.CLEAN_AREA,
+            name="Clean area",
```
```suggestion
            name="Cleaning Area",
```

frenck commented on homeassistant/components/tuya/sensor.py:
```diff
@@ -594,6 +594,24 @@ class TuyaSensorEntityDescription(SensorEntityDescription):
             subkey="voltage",
         ),
     ),
+	# Robot Vacuum
```
Tabs vs spaces

```suggestion
    # Robot Vacuum
```

frenck commented on homeassistant/components/tuya/sensor.py:
```diff
@@ -594,6 +594,24 @@ class TuyaSensorEntityDescription(SensorEntityDescription):
         ),
     ),
+	# Robot Vacuum
+    # https://developer.tuya.com/en/docs/iot/fsd?id=K9gf487ck1tlo
```
Should we add support for the others too? Like `total_clean_area`, `total_clean_count`, etc?

frenck commented on homeassistant/components/tuya/sensor.py:
```diff
@@ -594,6 +594,24 @@ class TuyaSensorEntityDescription(SensorEntityDescription):
+	# Robot Vacuum
+    # https://developer.tuya.com/en/docs/iot/fsd?id=K9gf487ck1tlo
+    "sd": (
+        TuyaSensorEntityDescription(
```
Does this need a `state_class`?

frenck commented on homeassistant/components/tuya/sensor.py:
```diff
@@ -594,6 +594,24 @@ class TuyaSensorEntityDescription(SensorEntityDescription):
+        ),
+        TuyaSensorEntityDescription(
+            key=DPCode.CLEAN_TIME,
+            name="Clean time",
```
```suggestion
            name="Cleaning Time",
```

frenck commented on homeassistant/components/tuya/sensor.py:
```diff
@@ -594,6 +594,24 @@ class TuyaSensorEntityDescription(SensorEntityDescription):
+            key=DPCode.CLEAN_AREA,
+            name="Clean area",
+        ),
+        TuyaSensorEntityDescription(
```
Should this have a nice icon as well?

frenck commented on homeassistant/components/tuya/select.py:
```diff
@@ -218,6 +218,26 @@
+            entity_category=ENTITY_CATEGORY_CONFIG,
+        ),
+        SelectEntityDescription(
+            key=DPCode.DIRECTION_CONTROL,
```
I'm not sure if this should be implemented as a select entity, to be honest. It provides steering controls (left, right, and such) for the vacuum cleaner. I don't think that a `select` provides the best user experience for such controls?

frenck commented on homeassistant/components/tuya/select.py:
```diff
@@ -218,6 +218,26 @@
             device_class=DEVICE_CLASS_TUYA_FINGERBOT_MODE,
             entity_category=ENTITY_CATEGORY_CONFIG,
         ),
+    ),    # Robot Vacuum
+    # https://developer.tuya.com/en/docs/iot/fsd?id=K9gf487ck1tlo
```
```suggestion
    ),
    # Robot Vacuum
    # https://developer.tuya.com/en/docs/iot/fsd?id=K9gf487ck1tlo
```

frenck commented on homeassistant/components/tuya/select.py:
```diff
@@ -218,6 +218,26 @@
         ),
+    ),    # Robot Vacuum
+    # https://developer.tuya.com/en/docs/iot/fsd?id=K9gf487ck1tlo
+    "sd": (
```
Should we implement others too? Like `collection_mode` and `cistern`?

dougiteixeira committed:
Update const.py

dougiteixeira committed:
Update vacuum.py

dougiteixeira committed:
Update homeassistant/components/tuya/select.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

dougiteixeira committed:
Update homeassistant/components/tuya/select.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

dougiteixeira committed:
Update homeassistant/components/tuya/select.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

dougiteixeira committed:
Update homeassistant/components/tuya/select.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

dougiteixeira committed:
Update homeassistant/components/tuya/sensor.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

dougiteixeira committed:
Update homeassistant/components/tuya/sensor.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

dougiteixeira committed:
Update homeassistant/components/tuya/sensor.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

dougiteixeira committed:
Update homeassistant/components/tuya/sensor.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

dougiteixeira committed:
Add icons Clean Area and Time

dougiteixeira committed:
Add sensors

Add total clean area, total clean time, clean count, duster cloth life, edge brush life, filter life, roll brush life and fault code.

dougiteixeira committed:
Add state_class sensors vacuum

dougiteixeira committed:
Add DPCODE CISTERN and COLLECTION_MODE

dougiteixeira committed:
Add vacuum CISTERN and COLLECTION_MODE

Add vacuum Water Tank Adjustment and Dust Collection Mode

dougiteixeira committed:
Update sensor.py

dougiteixeira committed:
Update sensor.py

dougiteixeira commented:
I'll cancel this PR and open new ones with smaller changes for review, thanks for the help.
